From b7d8f0ff2ff076f255f1af557ad09371082f1b9e Mon Sep 17 00:00:00 2001
From: Martin Jansa <martin.jansa@lge.com>
Date: Fri, 15 Oct 2021 05:48:56 -0700
Subject: [PATCH] python-3.10 compatibility

* as https://docs.python.org/3/whatsnew/3.10.html#id2 says:
  Since Py_REFCNT() is changed to the inline static function, Py_REFCNT(obj) = new_refcnt must be replaced with Py_SET_REFCNT(obj, new_refcnt): see Py_SET_REFCNT() (available since Python 3.9). For backward compatibility, this macro can be used:

  #if PY_VERSION_HEX < 0x030900A4
  #  define Py_SET_REFCNT(obj, refcnt) ((Py_REFCNT(obj) = (refcnt)), (void)0)
  #endif
  (Contributed by Victor Stinner in bpo-39573.)

* fixes:
In file included from /jenkins/mjansa/build/ros/webos-foxy-kirkstone/tmp-glibc/work/qemux86-webos-linux/python3-pyproj/1.9.6-r0/recipe-sysroot/usr/include/python3.10/genobject.h:12,
                 from /jenkins/mjansa/build/ros/webos-foxy-kirkstone/tmp-glibc/work/qemux86-webos-linux/python3-pyproj/1.9.6-r0/recipe-sysroot/usr/include/python3.10/Python.h:122,
                 from _proj.c:4:
/jenkins/mjansa/build/ros/webos-foxy-kirkstone/tmp-glibc/work/qemux86-webos-linux/python3-pyproj/1.9.6-r0/recipe-sysroot/usr/include/python3.10/abstract.h:354:17: note: declared here
  354 | PyAPI_FUNC(int) PyObject_AsWriteBuffer(PyObject *obj,
      |                 ^~~~~~~~~~~~~~~~~~~~~~
_proj.c: In function '__pyx_tp_dealloc_5_proj_Proj':
_proj.c:7991:5: error: lvalue required as increment operand
 7991 |     ++Py_REFCNT(o);
      |     ^~
_proj.c:7993:5: error: lvalue required as decrement operand
 7993 |     --Py_REFCNT(o);
      |     ^~

Upstream-Status: Pending

Signed-off-by: Martin Jansa <martin.jansa@lge.com>
---
 _proj.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/_proj.c b/_proj.c
index e3c7423..13f3ce7 100644
--- a/_proj.c
+++ b/_proj.c
@@ -7988,9 +7988,15 @@ static void __pyx_tp_dealloc_5_proj_Proj(PyObject *o) {
   {
     PyObject *etype, *eval, *etb;
     PyErr_Fetch(&etype, &eval, &etb);
-    ++Py_REFCNT(o);
+#if PY_VERSION_HEX < 0x030900A4
+#  define Py_SET_REFCNT(obj, refcnt) ((Py_REFCNT(obj) = (refcnt)), (void)0)
+#endif
+// maybe use
+// https://docs.python.org/3/c-api/refcounting.html#c.Py_INCREF
+// https://docs.python.org/3/c-api/refcounting.html#c.Py_DECREF
+    Py_SET_REFCNT(o, Py_REFCNT(o) + 1);
     __pyx_pw_5_proj_4Proj_3__dealloc__(o);
-    --Py_REFCNT(o);
+    Py_SET_REFCNT(o, Py_REFCNT(o) - 1);
     PyErr_Restore(etype, eval, etb);
   }
   Py_CLEAR(p->proj_version);
